Prepare the <model> for an immersive presentation

jeanhbr opened the pull request:
#### 77b3be1d7a731a6280de28e4637392c8ec4f7fb2
<pre>
Prepare the &lt;model&gt; for an immersive presentation
<a href="https://bugs.webkit.org/show_bug.cgi?id=303238">https://bugs.webkit.org/show_bug.cgi?id=303238</a>
<a href="https://rdar.apple.com/165527010">rdar://165527010</a>

Reviewed by Etienne Segonzac.

Add logic to ensure that a model player is running before entering the model in immersive
Add logic to allow the creation of the model player for hidden inline models
Remove the inline model from the rendered page

Test: model-element/immersive/model-element-immersive-hidden-inline.html

* LayoutTests/model-element/immersive/model-element-immersive-basic-expected.txt:
* LayoutTests/model-element/immersive/model-element-immersive-basic.html:
* LayoutTests/model-element/immersive/model-element-immersive-hidden-inline-expected.txt: Added.
* LayoutTests/model-element/immersive/model-element-immersive-hidden-inline.html: Copied from LayoutTests/model-element/immersive/model-element-immersive-no-client-approval.html.
* LayoutTests/model-element/immersive/model-element-immersive-no-client-approval.html:
Update test results to reflect the current state.
Add a test for hidden inline models.

* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
(WebCore::HTMLModelElement::setSourceURL):
(WebCore::HTMLModelElement::isVisible const):
(WebCore::HTMLModelElement::modelDidChange):
(WebCore::HTMLModelElement::createModelPlayer):
(WebCore::HTMLModelElement::reloadModelPlayer):
(WebCore::HTMLModelElement::configureGraphicsLayer):
(WebCore::HTMLModelElement::setEntityTransform):
(WebCore::HTMLModelElement::modelResourceFinished):
(WebCore::HTMLModelElement::ensureImmersivePresentation):
(WebCore::HTMLModelElement::exitImmersivePresentation):
(WebCore::HTMLModelElement::setDetachedForImmersive):
(WebCore::HTMLModelElement::ensureModelPlayer):
(WebCore::HTMLModelElement::triggerModelPlayerCreationCallbacksIfNeeded):
(WebCore::HTMLModelElement::modelContainerSizeIsEmpty const):
(WebCore::HTMLModelElement::sourceRequestResource):
Add a method to ensure the presence of a model player.
This will either directly give the current running model player if it exists,
wake up a suspended model player, or start a new model player creation request.

* Source/WebCore/Modules/model-element/HTMLModelElement.h:
* Source/WebCore/Modules/model-element/ModelPlayerGraphicsLayerConfiguration.h:
* Source/WebCore/dom/DocumentImmersive.cpp:
(WebCore::DocumentImmersive::exitImmersive):
(WebCore::DocumentImmersive::requestImmersive):
* Source/WebCore/dom/DocumentImmersive.h:

* Source/WebCore/platform/graphics/GraphicsLayer.h:
(WebCore::GraphicsLayer::removeModelContents):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::removeModelContents):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.cpp:
(WebKit::ModelProcessModelPlayer::configureGraphicsLayer):
Remove the model inline presentation if needed.

Canonical link: <a href="https://commits.webkit.org/303751@main">https://commits.webkit.org/303751@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c51ac431fda6eac88ecbc7b8b6f1f0ce699f0fc2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133505 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/6007 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/44651 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/141061 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/85556 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/135375 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/6521 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/5871 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102125 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/85556 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/136452 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/4614 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/119676 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/82923 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/4492 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2097 "Passed tests") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/113616 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/37786 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/143708 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/5676 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/38371 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/110501 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/5758 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/4865 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110683 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28056 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4358 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/115933 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/59426 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/5731 "Built successfully") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/5578 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/69183 "Built successfully") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/5820 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/5687 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -643,6 +657,11 @@ void HTMLModelElement::sizeMayHaveChanged()
 {
+#if ENABLE(MODEL_ELEMENT_IMMERSIVE)
+    if (!m_modelRendersInline)
+        return graphicsLayer.removeModelContents();
```
Rather than doing this directly, this would make more sense to do via the ModelPlayer, as each player may have a different notion of how the graphics layer is configured.

Could this bit, m_modelRendersInline, be added to the ModelPlayerGraphicsLayerConfiguration struct and passed to configureGraphicsLayer instead?

jeanhbr replied:
Indeed, thanks !

etiennesegonzac commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -472,9 +485,8 @@ void HTMLModelElement::createModelPlayer()
-    if (size.isEmpty())
-        return;
+    if (modelContainerSizeIsZero())
+        return didFinishModelPlayerCreation(Exception { ExceptionCode::AbortError, "Model container size is null"_s });
```
nit: "Model content size is empty" to match the release log info for the same condition. or the other way around :)

etiennesegonzac commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -1377,8 +1395,66 @@ void HTMLModelElement::requestImmersive(DOMPromiseDeferred<void>&& promise)
+void HTMLModelElement::ensureImmersivePresentation(CompletionHandler<void(ExceptionOr<const LayerHostingContextIdentifier>)>&& completion)
+{
+    setModelRendersInline(false);
+    ensureModelPlayer([weakThis = WeakPtr { *this }, completion = WTFMove(completion)] (auto result) mutable {
```
do we need weakThis? (and if so, should we check it)

etiennesegonzac commented on Source/WebCore/Modules/model-element/HTMLModelElement.h:
```diff
@@ -346,6 +349,16 @@ class HTMLModelElement final : public HTMLElement, private CachedRawResourceClie
+
+#if ENABLE(MODEL_ELEMENT_IMMERSIVE)
+    bool m_modelRendersInline { true };
+    void setModelRendersInline(bool);
```
Up to you, but I'm starting to think the code would be more readable with this flipped.
(m_modelRendersImmersive)

etiennesegonzac commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -1377,8 +1395,66 @@ void HTMLModelElement::requestImmersive(DOMPromiseDeferred<void>&& promise)
+
 #endif
 
+void HTMLModelElement::didFinishModelPlayerCreation(ExceptionOr<RefPtr<ModelPlayer>>&& result)
```
`triggerModelPlayerCreationCallbacksIfNeeded`?
(most of the call sites are not really didFinish)

etiennesegonzac commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -1377,8 +1395,66 @@ void HTMLModelElement::requestImmersive(DOMPromiseDeferred<void>&& promise)
+    for (auto& callback : std::exchange(m_modelPlayerCreationCallbacks, { }))
+        callback(result);
+#else
+        UNUSED_PARAM(result);
```
nit: extra indent?

jeanhbr committed:
Prepare the <model> for an immersive presentation
https://bugs.webkit.org/show_bug.cgi?id=303238
rdar://165527010

Reviewed by Etienne Segonzac.

Add logic to ensure that a model player is running before entering the model in immersive
Add logic to allow the creation of the model player for hidden inline models
Remove the inline model from the rendered page

Test: model-element/immersive/model-element-immersive-hidden-inline.html

* LayoutTests/model-element/immersive/model-element-immersive-basic-expected.txt:
* LayoutTests/model-element/immersive/model-element-immersive-basic.html:
* LayoutTests/model-element/immersive/model-element-immersive-hidden-inline-expected.txt: Added.
* LayoutTests/model-element/immersive/model-element-immersive-hidden-inline.html: Copied from LayoutTests/model-element/immersive/model-element-immersive-no-client-approval.html.
* LayoutTests/model-element/immersive/model-element-immersive-no-client-approval.html:
Update test results to reflect the current state.
Add a test for hidden inline models.

* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
(WebCore::HTMLModelElement::setSourceURL):
(WebCore::HTMLModelElement::isVisible const):
(WebCore::HTMLModelElement::modelDidChange):
(WebCore::HTMLModelElement::createModelPlayer):
(WebCore::HTMLModelElement::reloadModelPlayer):
(WebCore::HTMLModelElement::configureGraphicsLayer):
(WebCore::HTMLModelElement::setEntityTransform):
(WebCore::HTMLModelElement::modelResourceFinished):
(WebCore::HTMLModelElement::ensureImmersivePresentation):
(WebCore::HTMLModelElement::exitImmersivePresentation):
(WebCore::HTMLModelElement::setDetachedForImmersive):
(WebCore::HTMLModelElement::ensureModelPlayer):
(WebCore::HTMLModelElement::triggerModelPlayerCreationCallbacksIfNeeded):
(WebCore::HTMLModelElement::modelContainerSizeIsEmpty const):
(WebCore::HTMLModelElement::sourceRequestResource):
Add a method to ensure the presence of a model player.
This will either directly give the current running model player if it exists,
wake up a suspended model player, or start a new model player creation request.

* Source/WebCore/Modules/model-element/HTMLModelElement.h:
* Source/WebCore/Modules/model-element/ModelPlayerGraphicsLayerConfiguration.h:
* Source/WebCore/dom/DocumentImmersive.cpp:
(WebCore::DocumentImmersive::exitImmersive):
(WebCore::DocumentImmersive::requestImmersive):
* Source/WebCore/dom/DocumentImmersive.h:

* Source/WebCore/platform/graphics/GraphicsLayer.h:
(WebCore::GraphicsLayer::removeModelContents):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::removeModelContents):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.cpp:
(WebKit::ModelProcessModelPlayer::configureGraphicsLayer):
Remove the model inline presentation if needed.

Canonical link: https://commits.webkit.org/303751@main
